benchmark/multithread.cpp

```diff
@@ -0,0 +1,110 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include "multithread.hpp"
+
+#include <umf/memory_pool.h>
+#include <umf/pools/pool_jemalloc.h>
+#include <umf/pools/pool_scalable.h>
+#include <umf/providers/provider_os_memory.h>
+
+#include <iostream>
+#include <memory>
+#include <numeric>
+
+static constexpr size_t N_REPEATS = 5;
+static constexpr size_t N_ITERATIONS = 50000;
+static constexpr size_t N_THREADS = 20;
+static constexpr size_t ALLOC_SIZE = 64;
+
+using poolCreateExtParams = std::tuple<umf_memory_pool_ops_t *, void *,
+                                       umf_memory_provider_ops_t *, void *>;
+
+static auto poolCreateExtUnique(poolCreateExtParams params) {
+    umf_memory_pool_handle_t hPool;
+    auto [pool_ops, pool_params, provider_ops, provider_params] = params;
+
+    umf_memory_provider_handle_t provider = nullptr;
+    auto ret =
+        umfMemoryProviderCreate(provider_ops, provider_params, &provider);
+    if (ret != UMF_RESULT_SUCCESS) {
+        std::cerr << "provider create failed" << std::endl;
+        abort();
+    }
+
+    ret = umfPoolCreate(pool_ops, provider, pool_params,
+                        UMF_POOL_CREATE_FLAG_OWN_PROVIDER, &hPool);
+    if (ret != UMF_RESULT_SUCCESS) {
+        std::cerr << "pool create failed" << std::endl;
+        abort();
+    }
+
+    return std::shared_ptr<umf_memory_pool_t>(hPool, &umfPoolDestroy);
+}
+
+static void mt_alloc_free(poolCreateExtParams params) {
+    auto pool = poolCreateExtUnique(params);
+
+    std::vector<void *> allocs[N_THREADS];
+    size_t numFailures[N_THREADS] = {};
+    for (auto &v : allocs) {
+        v.reserve(N_ITERATIONS);
+    }
+
+    auto values = umf_bench::measure<std::chrono::milliseconds>(
+        N_REPEATS, N_THREADS, [&, pool = pool.get()](auto thread_id) {
+            for (int i = 0; i < N_ITERATIONS; i++) {
+                allocs[thread_id].push_back(umfPoolMalloc(pool, ALLOC_SIZE));
+                if (!allocs[thread_id].back()) {
+                    numFailures[thread_id]++;
+                }
+            }
+
+            for (int i = 0; i < N_ITERATIONS; i++) {
+                umfPoolFree(pool, allocs[thread_id][i]);
+            }
+
+            // clear the vector as this function might be called multiple times
+            allocs[thread_id].clear();
+        });
+
+    std::cout << "mean: " << umf_bench::mean(values)
```

Comment:
I think that the output in one line is more clear:
```
scalable_pool mt_alloc_free: mean: 29.27 [ms] std_dev: 22.1912 [ms] (total alloc failures: 0 out of 5000000)
jemalloc_pool mt_alloc_free: mean: 431.32 [ms] std_dev: 43.2909 [ms] (total alloc failures: 0 out of 5000000)
```
than the current output in 3 lines:
```
scalable_pool mt_alloc_free: 
mean: 29.27 [ms] std_dev: 22.1912 [ms]
Total alloc failures: 0 out of [5](https://github.com/oneapi-src/unified-memory-framework/actions/runs/7789245292/job/21240602362?pr=204#step:7:6)000000
jemalloc_pool mt_alloc_free: 
mean: 431.32 [ms] std_dev: 43.2[9](https://github.com/oneapi-src/unified-memory-framework/actions/runs/7789245292/job/21240602362?pr=204#step:7:10)09 [ms]
Total alloc failures: 0 out of 5000000
```

Reply:
Done.